calico_rkt/calico_policy.py

```diff
@@ -0,0 +1,156 @@
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from constants import *
+from calico_rkt import configure_logging
+from pycalico.datastore import DatastoreClient
+from pycalico.datastore_datatypes import Rule, Rules
+from pycalico.utils import validate_characters
+
+KUBERNETES_DEFAULT_PROFILE = "default_profile"
+
+_log = logging.getLogger(__name__)
+configure_logging(_log)
+
+class BasePolicyDriver(object):
+    """
+    Abstract base class of a CNI Policy Driver.
+
+    The CNI Policy Driver is responsible for applying network policy to
+    networked containers and creating/applying Calico profiles to Calico
+    endpoints.
+    """
+    def __init__(self):
+
+        self._client = DatastoreClient()
+        """
+        DatastoreClient for access to the Calico datastore.
+        """
+
+        self.profile_name = None
+        """
+        Name of profile for attach to endpoint. Must be set in init function
+        of subclass.
+        """
+
+    def set_profile(self, endpoint):
+        """Sets a profile for the networked container on the given endpoint.
+
+        Create a profile if it is not yet created.
+
+        :param endpoint:
+        :return: None
+        """
+        assert self.profile_name, "No profile name set."
+        if not self._client.profile_exists(self.profile_name):
+            # If the profile doesn't exist, create it.
+            _log.info("Creating new profile '%s'", self.profile_name)
+            rules = self.generate_rules()
+            self._client.create_profile(self.profile_name, rules)
+
+        # Set the default profile on this pod's Calico endpoint.
+        _log.info("Setting profile '%s' on endpoint %s",
+                  self.profile_name, endpoint.endpoint_id)
+        self._client.set_profiles_on_endpoint(profile_names=[self.profile_name],
+                                              endpoint_id=endpoint.endpoint_id)
+
+    def remove_profile(self):
+        """Remove the profile if there are no endpoints attached.
+
+        :return: None
+        """
+        raise NotImplementedError("Must implement method in subclass")
+
+    def generate_rules(self):
+        """Generates Calico Rules
+
+        :rtype: Calico Rules datatype or None
+        :return: rules - contains inbound and outbound policy rules; None
+        uses default rules
+        """
+        raise NotImplementedError("Must implement method in subclass")
+
+
+class DefaultPolicyDriver(BasePolicyDriver):
+    """
+    Implements default network policy for a generic CNI plugin.
+    """
+    def __init__(self, network_name):
+        BasePolicyDriver.__init__(self)
+        if not validate_characters(network_name):
+            raise ValueError("Invalid characters detected in the given network "
```

Comment:
Can we print a list of valid characters so the user knows what is wrong? Is that defined in libcalico somewhere?